### ISSUE TITLE
head() is slower than take()

### ISSUE THREAD

**Issue by wjones127** [lance-format/lance#1927]
We aren't pushing the limit down enough in the plan. We should either do that or just convert head into take when appropriate.

**Comment by jayzhan211** [comment]
what is the head() and take() referred to?

Is it API in this code?
```py
def test_get_fragments(tmp_path: Path):
    table = pa.Table.from_pydict({"a": range(100), "b": range(100)})
    base_dir = tmp_path / "test"
    lance.write_dataset(table, base_dir)

    dataset = lance.dataset(base_dir)
    fragment = dataset.get_fragments()[0]
    assert fragment.count_rows() == 100
    assert fragment.physical_rows == 100
    assert fragment.num_deletions == 0

    assert fragment.metadata.id == 0

    head = fragment.head(10)
    tm.assert_frame_equal(head.to_pandas(), table.to_pandas()[0:10])

    assert fragment.to_table() == table

    taken = fragment.take([18, 20, 33, 53])
    assert taken == pa.Table.from_pydict({"a": [18, 20, 33, 53], "b": [18, 20, 33, 53]})
```

And, we would like to replace `head(n)` with `take([0, .., n -1])`?

**Comment by wjones127** [comment]
> And, we would like to replace head(n) with take([0, .., n -1])?

Yes, basically what we want. This should happen at the rust level. Right now it uses `ScanExec` to get the head, but that does too much readahead. In the `Scanner::create_plan()` method we should use some other exec node instead of `ScanExec` when we are just doing head. `TakeExec` might not be exactly what we want, since it expects an input. Could create another node that wraps `Dataset::take()`.

**Comment by jayzhan211** [comment]
I'm thinking of replacing LanceScanExec in `scan_fragments` with a new ExecutionPlan that wraps `Dataset::take()`. Does this make sense?

**Comment by wjones127** [comment]
> Does this make sense?

Sure, that's a good place to start.

**Comment by Xuanwo** [comment]
Resolved by #2747, which pushes limit and offset into the file reader for scans without filters. Dataset.head now reads the requested range instead of relying on a terminal limit after excessive readahead. Closing as fixed.
